### ISSUE TITLE
SSH login sometimes fails

### ISSUE THREAD

**Issue by AlexanderPavlenko** [romanz/trezor-agent#224]
```
> trezor-agent --debug -v -s ubuntu@somehost
2018-03-24 12:54:21,089 INFO         identity #0: <ssh://ubuntu@somehost|nist256p1>                                        [__init__.py:251]
2018-03-24 12:54:21,103 WARNING      DISPLAY not defined                                                                                  [ui.py:65]
2018-03-24 12:54:21,103 INFO         using [b'ttyname=/dev/ttys002'] for pinentry options                                                 [ui.py:67]
2018-03-24 12:54:21,147 INFO         running '/bin/zsh' with {'SSH_AUTH_SOCK': '/var/folders/g0/.../T/trezor-ssh-agent-_8q6_pge', 'SSH_AGENT_PID': '66729'} [server.py:119]

> ssh ubuntu@somehost
2018-03-24 12:54:40,172 INFO         please confirm user "ubuntu" login to "<ssh://ubuntu@somehost|nist256p1>" using Trezor... [client.py:46]
2018-03-24 12:54:42,890 INFO         signature status: OK                                                                                 [protocol.py:151]
# connected

> ssh ubuntu@somehost
2018-03-24 12:54:46,935 INFO         please confirm user "ubuntu" login to "<ssh://ubuntu@somehost|nist256p1>" using Trezor... [client.py:46]
2018-03-24 12:54:49,706 INFO         signature status: OK                                                                                 [protocol.py:151]
# connected

> ssh ubuntu@somehost
2018-03-24 12:54:57,933 INFO         please confirm user "ubuntu" login to "<ssh://ubuntu@somehost|nist256p1>" using Trezor... [client.py:46]
sign_and_send_pubkey: signing failed: communication with agent failed
[1]    66729 illegal hardware instruction  trezor-agent --debug -v -s ubuntu@somehost

ubuntu@somehost: Permission denied (publickey).
zsh: error on TTY read: Input/output error
# terminal window closes, OS says "python exited unexpectedly"
```

macOS 10.13.3
Trezor T 2.0.6
Python 3.6.4

Please let me know if there's any additional info needed.

**Comment by romanz** [comment]
Thanks for reporting this!
It seems related to #157 (Python segfaulting on Mac OS X) - but since I don't have an Mac, it would be hard for me to debug this issue.

I guess that this failure may be related to USB-related [library](https://github.com/trezor/python-trezor/blob/e1e419485fc9247918b9a15d9272fdc7b1fdadea/trezorlib/transport/webusb.py#L22) (used by the `python-trezor` library to communicate with the TREZOR model T device) having an issue on Mac OS...

@galuszkak @prusnak What do you think?

**Comment by prusnak** [comment]
@AlexanderPavlenko How did you install python-trezor dependencies? If anything was installed bottled brew package, I would say the problem is in incorrect build flags. Try to build all python-trezor dependencies from source to see if that fixed the problem.

**Comment by AlexanderPavlenko** [comment]
It is installed via pip.

```
> pip --version
pip 9.0.3 from /usr/local/lib/python3.6/site-packages (python 3.6)

> pip install --force-reinstall -U trezor-agent
Collecting trezor-agent
Collecting trezor>=0.9.0 (from trezor-agent)
Collecting libagent>=0.9.0 (from trezor-agent)
Collecting rlp>=0.6.0 (from trezor>=0.9.0->trezor-agent)
Collecting hidapi>=0.7.99.post20 (from trezor>=0.9.0->trezor-agent)
Collecting requests>=2.4.0 (from trezor>=0.9.0->trezor-agent)
  Using cached requests-2.18.4-py2.py3-none-any.whl
Collecting libusb1>=1.6.4 (from trezor>=0.9.0->trezor-agent)
Collecting mnemonic>=0.17 (from trezor>=0.9.0->trezor-agent)
Collecting setuptools>=19.0 (from trezor>=0.9.0->trezor-agent)
  Using cached setuptools-39.0.1-py2.py3-none-any.whl
Collecting ecdsa>=0.9 (from trezor>=0.9.0->trezor-agent)
  Using cached ecdsa-0.13-py2.py3-none-any.whl
Collecting pyblake2>=0.9.3 (from trezor>=0.9.0->trezor-agent)
Collecting click>=6.2 (from trezor>=0.9.0->trezor-agent)
  Using cached click-6.7-py2.py3-none-any.whl
Collecting semver>=2.2 (from libagent>=0.9.0->trezor-agent)
Collecting ed25519>=1.4 (from libagent>=0.9.0->trezor-agent)
Collecting unidecode>=0.4.20 (from libagent>=0.9.0->trezor-agent)
  Using cached Unidecode-1.0.22-py2.py3-none-any.whl
Collecting backports.shutil-which>=3.5.1 (from libagent>=0.9.0->trezor-agent)
  Using cached backports.shutil_which-3.5.1-py2.py3-none-any.whl
Collecting pymsgbox>=1.0.6 (from libagent>=0.9.0->trezor-agent)
Collecting python-daemon>=2.1.2 (from libagent>=0.9.0->trezor-agent)
  Using cached python_daemon-2.1.2-py2.py3-none-any.whl
Collecting ConfigArgParse>=0.12.0 (from libagent>=0.9.0->trezor-agent)
Collecting urllib3<1.23,>=1.21.1 (from requests>=2.4.0->trezor>=0.9.0->trezor-agent)
  Using cached urllib3-1.22-py2.py3-none-any.whl
Collecting chardet<3.1.0,>=3.0.2 (from requests>=2.4.0->trezor>=0.9.0->trezor-agent)
  Using cached chardet-3.0.4-py2.py3-none-any.whl
Collecting certifi>=2017.4.17 (from requests>=2.4.0->trezor>=0.9.0->trezor-agent)
  Using cached certifi-2018.1.18-py2.py3-none-any.whl
Collecting idna<2.7,>=2.5 (from requests>=2.4.0->trezor>=0.9.0->trezor-agent)
  Using cached idna-2.6-py2.py3-none-any.whl
Collecting pbkdf2 (from mnemonic>=0.17->trezor>=0.9.0->trezor-agent)
Collecting docutils (from python-daemon>=2.1.2->libagent>=0.9.0->trezor-agent)
  Using cached docutils-0.14-py3-none-any.whl
Collecting lockfile>=0.10 (from python-daemon>=2.1.2->libagent>=0.9.0->trezor-agent)
  Using cached lockfile-0.12.2-py2.py3-none-any.whl
```

> Try to build all python-trezor dependencies from source

How this can be done?

**Comment by Quaternioneer** [comment]
2018-04-13 15:00:31,440 INFO         identity #0: <ssh://anonymous@hostname.net|nist256p1>                                       [__init__.py:251]
2018-04-13 15:00:31,442 INFO         using [b'ttyname=/dev/pts/5', b'display=:0'] for pinentry options                                    [ui.py:67]
2018-04-13 15:00:31,445 INFO         running ['ssh', '-l', 'anonymous', 'hostname.net'] with {'SSH_AUTH_SOCK': '/tmp/trezor-ssh-agent-c0ndeo6m', 'SSH_AGENT_PID': '7623'} [server.py:119]
2018-04-13 15:00:31,720 INFO         Starting new HTTP connection (1): 127.0.0.1                                                          [connectionpool.py:208]
2018-04-13 15:00:31,730 INFO         Starting new HTTP connection (1): 127.0.0.1                                                          [connectionpool.py:208]

My ledger-agent works, but trezor-agent does not.  I understand environment variables and virtual enviornments only vaguely; my ledger-agent is installed under python3.5 and I now pip3 defaults to python3.6.  However, I briefly reinstalled trezor-agent under python3.5 without any success.  Unfortunately, I don't have an original trezor to try with trezor-agent.  Ledger-agent and trezor-agent only differ in the last two lines of the above paragraph of output. Ledger says,
2018-04-13 15:14:40,126 INFO         please confirm user "anonymous" login to "<ssh://anonymous@hostname|nist256p1>" using LedgerNanoS... [client.py:46]

Ubuntu 16 LTS 
Model T firmware updated today

**Comment by Quaternioneer** [comment]
Somehow, I installed Ledger agent under /home/.../.local instead of Unix system resources. The Ledger and Trezor agents are both installed under  /usr/local/lib/python3.6/dist-packages now.  Ledger agent still works and Trezor agent still does not.

**Comment by romanz** [comment]
It seems that `trezor-agent` doesn't succeed in connecting to the TREZOR device (but instead tries to connect via the Bridge).
Could you run `trezor-agent foo -vv` and attach the logs (to debug the connection issue)?
The output on my machine looks like:
```
trezor-agent i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE -vv -e nist256p1
2018-04-14 11:44:52,735 DEBUG        parsed identity: {'proto': None, 'user': None, 'host': 'i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE', 'port': None, 'path': None} [interface.py:30]
2018-04-14 11:44:52,735 DEBUG        identity parts: ['ssh://', 'i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE']                                       [interface.py:46]
2018-04-14 11:44:52,735 INFO         identity #0: <ssh://i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE|nist256p1>                                      [__init__.py:251]
2018-04-14 11:44:52,737 INFO         using [b'ttyname=/dev/pts/2', b'display=:0.0'] for pinentry options                                  [ui.py:67]
2018-04-14 11:44:52,826 DEBUG        Starting new HTTP connection (1): 127.0.0.1                                                          [connectionpool.py:208]
2018-04-14 11:44:52,833 DEBUG        transports: [<trezorlib.transport.webusb.WebUsbTransport object at 0x7f8a575d1d68>]                  [trezor.py:111]
2018-04-14 11:44:53,518 DEBUG        connected to Trezor <snip>                                                         [trezor.py:91]
2018-04-14 11:44:53,519 DEBUG        vendor   : trezor.io                                                                                 [trezor.py:93]
2018-04-14 11:44:53,519 DEBUG        version  : 2.0.6                                                                                     [trezor.py:97]
2018-04-14 11:44:53,550 DEBUG        identity parts: ['ssh://', 'i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE']                                       [interface.py:46]
2018-04-14 11:44:53,551 DEBUG        "<ssh://i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE|nist256p1>" getting public key (nist256p1) from Trezor      [trezor.py:139]
2018-04-14 11:44:53,551 DEBUG        identity parts: ['ssh://', 'i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE']                                       [interface.py:46]
2018-04-14 11:44:53,551 DEBUG        bip32 address string: b'\x00\x00\x00\x00ssh://i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE'                      [interface.py:88]
2018-04-14 11:44:53,674 DEBUG        result: <snip> [trezor.py:143]
2018-04-14 11:44:53,707 DEBUG        identity parts: ['ssh://', 'i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE']                                       [interface.py:46]
2018-04-14 11:44:53,707 DEBUG        fingerprint: <snip>                                         [formats.py:189]
ecdsa-sha2-nistp256 <snip> <ssh://i3C4L4fvmJejQ5aatWTsYaxaf0DRqIcE|nist256p1>
```

**Comment by Quaternioneer** [comment]
I commented out my udev rules and reloaded, but that did not seem to change the debug log.
[trezor.log](https://github.com/romanz/trezor-agent/files/1910696/trezor.log)

**Comment by Quaternioneer** [comment]
I had 2.0.12.  I reinstalled Ubuntu to eliminate the possibility that I misconfigured something smart card related when I set up Ledger agent and Yubikey.  I cannot reproduce the error at the moment.  If I recall correctly, neither of these variables were set, SSH_AUTH_SOCK and GPG_AGENT_INFO.  In my new environment, they now equal,
/home/nebulay/.gnupg/S.gpg-agent.ssh and
/home/nebulay/.gnupg/S.gpg-agent:0:1 respectively. 

Perhaps I could help with documentation. It took me a couple weeks to realize that Trezor agent responds with a different public key for each user@hostname.  I expected behavior similar to Yubikey.
